### PR TITLE
ControlledGate implementation

### LINKED ISSUE #236: Add support for arbitrary controlled gates
We want to be able to generate a [W-state](https://en.wikipedia.org/wiki/W_state), and for that, we need a controlled-Hadamard gate.

Since a Hadamard gate is dense, it makes sense to go ahead and make a generic controlled gate class.

### PR DESCRIPTION
**Pull request by fSosaRey**
Kernel is defined using DenseOperator{2}, and apply_gate dispatches to specialized apply_controlled_gate!

We could add a specialization using a Diagonal kernel or an Antidiagonal kernel, but the Pauli gates fitting those descriptions are already optimized individually.

DenseOperators of arbitraty size, (multiple controls) could also be created, but again, Toffoli has its optimization. 

Closes #236

### CODE REVIEW

**Comment by fSosaRey** [comment]
also, are we happy with this printout?

```
julia> ControlledGate(hadamard(),2,3)
Gate Object: ControlledGate
Connected_qubits	: [2, 3]
Operator:
(4, 4)-element Snowflake.DenseOperator:
Underlying data ComplexF64:
1.0 + 0.0im    0.0 + 0.0im    0.0 + 0.0im    0.0 + 0.0im
0.0 + 0.0im    1.0 + 0.0im    0.0 + 0.0im    0.0 + 0.0im
0.0 + 0.0im    0.0 + 0.0im    0.7071067811865475 + 0.0im    0.7071067811865475 + 0.0im
0.0 + 0.0im    0.0 + 0.0im    0.7071067811865475 + 0.0im    -0.7071067811865475 + 0.0im
```

**Comment by plerouxanyon** [comment]
> also, are we happy with this printout?
> 
> ```
> julia> ControlledGate(hadamard(),2,3)
> Gate Object: ControlledGate
> Connected_qubits	: [2, 3]
> Operator:
> (4, 4)-element Snowflake.DenseOperator:
> Underlying data ComplexF64:
> 1.0 + 0.0im    0.0 + 0.0im    0.0 + 0.0im    0.0 + 0.0im
> 0.0 + 0.0im    1.0 + 0.0im    0.0 + 0.0im    0.0 + 0.0im
> 0.0 + 0.0im    0.0 + 0.0im    0.7071067811865475 + 0.0im    0.7071067811865475 + 0.0im
> 0.0 + 0.0im    0.0 + 0.0im    0.7071067811865475 + 0.0im    -0.7071067811865475 + 0.0im
> ```

Yes, I'm happy with that. The printout is a minor detail and super easy to change.

**Comment by plerouxanyon** [comment]
> @plerouxanyon should we convert ControlX and ControlZ gates to ControlledGates? there would be a performance gain, as right now they use DenseOperator{4}, although only the 2x2 kernel has an effect

Yes, but let's do that in the next PR to prevent scope creep in this one.

File: src/core/quantum_gate.jl [hunk 1]
```diff
@@ -101,6 +133,106 @@ get_connected_qubits(gate::AbstractGate)=[gate.target]
...
+   qubit_count: 2 
+q[1]:──*──
+       |  
+q[2]:──H──
+          
+```
+
+
+"""
+struct ControlledGate{G<:AbstractGate}<:AbstractControlledGate
```
Comment (plerouxanyon) [inline src/core/quantum_gate.jl:178]: I think we should remove `AbstractControlledGate` since we know a gate is a controlled gate if it is this type.

File: src/core/quantum_gate.jl [hunk 2]
```diff
@@ -101,6 +133,106 @@ get_connected_qubits(gate::AbstractGate)=[gate.target]
...
+
+        symbol_specs=gates_display_symbols[kernel_gate_type]      
+        num_params=length(params)
+        gate_params=Dict{String,Real}() 
+        
+        for (field,param) in zip(symbol_specs[end-num_params+1:end],params)
+            gate_params[field]=param
+        end
+
+        gates_display_symbols[ControlledGate{kernel_gate_type}]=[
```
Comment (plerouxanyon) [inline src/core/quantum_gate.jl:211]: Mutating `gates_display_symbols` is hard to follow and non-obvious for someone not familiar with Snowflake.

If someone wanted to update how the `ControlledGate{XGate}` rendered, for example. They would look at `Base.show` they would, follow along and eventually see that it looks inside `gates_display_symbols`, and would be very confused to see that `ControlledGate{XGate}` is not present. They might find this if they search for all places `gates_display_symbols` is defined, but this is not obvious, and mutable changes like this tend to become fragile over time.

A good mental check to do to verify if code is readable is to put yourself in the shoes of someone that has never seen this code before. How will they navigate this code?

We need to improve this at some point, but for now, let's simply have a `get_controlled_gate_symbol` function which looks at the inner gate and adds the `*` symbols.

File: src/core/quantum_gate.jl [hunk 3]
```diff
@@ -101,6 +133,106 @@ get_connected_qubits(gate::AbstractGate)=[gate.target]
...
+q[1]:──*──
+       |  
+q[2]:──H──
+          
+```
+
+
+"""
+struct ControlledGate{G<:AbstractGate}<:AbstractControlledGate
+    kernel::AbstractOperator
```
Comment (plerouxanyon) [inline src/core/quantum_gate.jl:179]: I think the kernel should be a gate (gate symbol after https://github.com/anyonlabs/Snowflake.jl/pull/232) instead of an operator.

EDIT: I understand why you made it an operator due to how the gate symbols are managed, but if we change that, then, to me, at least, it makes more sense to make it a gate.

File: src/core/quantum_gate.jl [hunk 4]
```diff
@@ -92,6 +98,32 @@ q[1]:──X45──
 q[2]:───────
 ```
 
+Lastly, to enable the construction of a `ControlledGate{X45}` made with this
```
Comment (plerouxanyon) [inline src/core/quantum_gate.jl:101]: If we update the display for controlled gates, this will fall away.

File: src/core/quantum_gate.jl [hunk 5]
```diff
@@ -101,6 +133,106 @@ get_connected_qubits(gate::AbstractGate)=[gate.target]
...
+                precision=precision
+            )[1]
+        ]
+        
+        new{kernel_gate_type}(kernel,connected_qubits)
+    end
+end
+
+# returns an equivalent size 4 DenseOperator built using the kernel
+function get_operator(gate::ControlledGate, T::Type{<:Complex}=ComplexF64)
```
Comment (plerouxanyon) [inline src/core/quantum_gate.jl:226]: This assumes a single-qubit single-control gate, but the definition of our gate is general. What if we have a controlled swap gate? Or a multi-controlled X gate (such as Tofolli).

File: src/core/quantum_gate.jl [hunk 6]
```diff
@@ -101,6 +133,106 @@ get_connected_qubits(gate::AbstractGate)=[gate.target]
...
+# returns an equivalent size 4 DenseOperator built using the kernel
+function get_operator(gate::ControlledGate, T::Type{<:Complex}=ComplexF64) 
+    op=get_matrix(eye(4))
+    
+    op[3:4,3:4]=get_matrix(gate.kernel)
+
+    return DenseOperator(convert(Matrix{T},op))
+end
+
+get_connected_qubits(gate::ControlledGate)=gate.connected_qubits
```
Comment (plerouxanyon) [inline src/core/quantum_gate.jl:234]: The gate will only contain the target qubits, but connected qubits should also include the control qubits.

File: test/circuit.jl [hunk 7]
```diff
@@ -226,3 +231,59 @@ end
...
+        ControlledGate( rotation,      connected_qubits; params=[pi,pi/2]),
+        ControlledGate( rotation_x,    connected_qubits; params=[pi/3]),
+        ControlledGate( rotation_y,    connected_qubits; params=[pi/4]),
+        ControlledGate( phase_shift,   connected_qubits; params=[pi/7]),
+        ControlledGate( universal,     connected_qubits; params=[pi/3,pi/12,pi/4]),
+    ]
+
+    circuit=QuantumCircuit(qubit_count=2,gates=gate_list)
+
+    print(circuit)
```
Comment (plerouxanyon) [inline test/circuit.jl:287]: This isn't really a test. All that this does is make sure it doesn't crash. It doesn't test that what it does is correct. I would recommend either removing this test or extending this test to ensure that the output is what is expected.

File: test/transpile.jl [hunk 8]
```diff
@@ -962,3 +962,11 @@ end
     @test !circuit_contains_gate_type(transpiled_circuit, Snowflake.Swap)
     @test simulate(circuit) ≈ simulate(transpiled_circuit)
 end
+
+@testset "CastControlledGatesToNativeTranspiler" begin
```
Comment (plerouxanyon) [inline test/transpile.jl:966]: Native gates depend on the QPU and shouldn't be part of the transpiler name. I think renaming it to `UnsupportedGatesTranspiler` would be more apt, or any other name that does not make assumptions about the QPU.

**Review by plerouxanyon — CHANGES_REQUESTED** [review]

File: src/core/quantum_gate.jl (hunk 2 above)
Comment (fSosaRey) [reply src/core/quantum_gate.jl:211]: @plerouxanyon do you mean that display symbol construction will be handled in `get_display_symbol(::ControlledGate)` and not stored in `gates_display_symbols`?

File: src/core/quantum_gate.jl (hunk 6 above)
Comment (fSosaRey) [reply src/core/quantum_gate.jl:234]: @plerouxanyon  the `connected_qubits` passed at construction of ControlledGate has both

File: src/core/quantum_gate.jl (hunk 2 above)
Comment (plerouxanyon) [reply src/core/quantum_gate.jl:211]: Yes, since all controlled gates will be handled the same way. The controlled qubits will all be `*`, and the rest will be the regular symbols from `get_display_symbols`.

File: src/core/quantum_gate.jl (hunk 1 above)
Comment (fSosaRey) [reply src/core/quantum_gate.jl:178]: updated to <:AbstractGate

File: src/core/quantum_gate.jl (hunk 5 above)
Comment (fSosaRey) [reply src/core/quantum_gate.jl:226]: ok the fall-back case for arbitrairy ControlledGates is to build a DenseOperator and apply that

File: test/circuit.jl (hunk 7 above)
Comment (fSosaRey) [reply test/circuit.jl]: this was removed, each ControlledGate is tested individually for producing the correct labels

File: test/transpile.jl (hunk 8 above)
Comment (fSosaRey) [reply test/transpile.jl]: name changed

File: src/core/quantum_gate.jl (hunk 4 above)
Comment (fSosaRey) [reply src/core/quantum_gate.jl]: removed

File: src/core/quantum_gate.jl (hunk 2 above)
Comment (fSosaRey) [reply src/core/quantum_gate.jl]: updated

**Review by plerouxanyon — APPROVED** [review]

File: src/core/quantum_gate.jl [hunk 9]
```diff
@@ -101,6 +126,124 @@ get_connected_qubits(gate::AbstractGate)=[gate.target]
...
+
+"""
+struct ControlledGate{GateType<:AbstractGate}<:AbstractGate
+    kernel::GateType
+    connected_qubits::Vector{Int}
+    control_qubits::Vector{Int}
+    kernel_qubits::Vector{Int}
+
+    function ControlledGate(
+        kernel_handle::Function,
```
Comment (plerouxanyon) [inline src/core/quantum_gate.jl:202]: Don't take a function that constructs a gate. Take an already constructed gate. This limits what the user can define as their own gate.

**Review by plerouxanyon — CHANGES_REQUESTED** [review]

File: src/core/quantum_gate.jl (hunk 9 above)
Comment (fSosaRey) [reply src/core/quantum_gate.jl:202]: ok, though this opens up the recursive ControlledGate{ControlledGate{...}}. Should I put safeguards against that?

Comment (fSosaRey) [reply src/core/quantum_gate.jl]: updated, with assertion error for ControlledGate{ControlledGate}

**Review by plerouxanyon — APPROVED** [review]
